### PR TITLE
Added support for multidimensional array indexing

### PR DESCRIPTION
**Pull request by kc611**
This PR builds on top of #8238 

As titled, this PR adds support for multidimensional indices while indexing NumPy arrays in Numba JIT functions: 

```python
import numba
import numpy as np

@numba.njit
def func(x, idx):
    return x[:, 2, idx, :, :]

a = np.random.randint(0, 100, (10, 11, 12, 13, 14))
b = np.random.randint(0, 10, (4, 5)) # Previously, these were limited to 1-D.

print(func(a, b).shape) 
# (10, 4, 5, 13, 14)
print(np.allclose(func(a, b), func.py_func(a, b)))
# True
```

### CODE REVIEW

**Comment by kc611** [comment]
The memory leak I was facing in this issue has been isolated: 

```python
import numpy as np
import unittest
from numba import njit
from numba.tests.support import MemoryLeakMixin, TestCase

class TestFancyIndexingMultiDim(MemoryLeakMixin, TestCase):
    shape = (4, 5, 6)
    
    def check_setitem_indices(self, arr_shape, index):
        @njit     
        def set_item(array, idx, item):
            array[idx] = item

        arr = np.random.randint(0, 11, size=arr_shape)
        src = arr[index]
        expected = np.zeros_like(arr)
        got = np.zeros_like(arr)

        set_item.py_func(expected, index, src)
        set_item(got, index, src)

        np.testing.assert_equal(got, expected)

    def test_setitem_with_tuple(self):
        # When index is an array within a tuple
        idx = (np.array([1,2,3]),)
        # No memory leak
        self.check_setitem_indices(self.shape, idx)

    def test_setitem_without_tuple(self):
        # When index itself is an array, it is
        # supposed to be implicitly cast to same form as
        # the example above i.e. both should and do 
        # give same answer, except for the leak
        idx = np.array([1,2,3])
        # Memory leaks
        self.check_setitem_indices(self.shape, idx)
    
if __name__ == '__main__':
    unittest.main()

```

**Comment by stuartarchibald** [comment]
@kc611 please could you resolve conflicts against `main`? As #8238 is merged the diff will hopefully shrink. Many thanks.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1711]: I don't think that this path in its present form can be supported on CUDA as the `flat_imp*` function compiled below relies on:
* NumPy functions
* CPU only implementations for `reshape`
* returns an array
* needs the NRT

I'm not sure of the best way to "detect" this, immediate thoughts are to either require the NRT and raise if the context doesn't have it, or alternatively declare these functions as overloads targetting CPU only which will then fail for CUDA users. CC @gmarkall do you have an opinion on this?

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1157]: As per a later comment, this _requires_ the presence of the NRT which isn't guaranteed and needs guarding against.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1662]: Given this check is removed, does the proposed implementation correctly handle "overlap"?

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:1]: It would be good for the testing to now also include indexing into F-order and A-order arrays?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the PR @kc611, great to see this implemented. I've given this an initial review and have also done some initial manual testing. Once the comments are address I'll look at the implementation details more closely, but from a cursory inspection the approach seems like it should work. Thanks again!

File: numba/core/typing/arraydecl.py
Comment (kc611) [reply numba/core/typing/arraydecl.py:89]: 'One-dimensional boolean indexing array per axis of source array / multidimensional boolean index array in a single axis of source array are not allowed.' is what it's trying to convey.

I'll add clarification in the `msg`

File: numba/np/arrayobj.py
Comment (kc611) [reply numba/np/arrayobj.py:769]: This is the flattened indexing array not the resultant array, it's used only for internal operations of `Indexer`s and discarded when the operation completes.

File: numba/np/arrayobj.py
Comment (kc611) [reply numba/np/arrayobj.py:786]: I'd say improbable, theoretically you'll need to have an array of size 2^32 to overflow in a 32-bit system.

File: numba/np/arrayobj.py
Comment (kc611) [reply numba/np/arrayobj.py:1662]: Define overlap in context of arrays ?

File: numba/tests/test_fancy_indexing.py
Comment (kc611) [reply numba/tests/test_fancy_indexing.py:1]: Alright will add that.

File: numba/np/arrayobj.py
Comment (kc611) [reply numba/np/arrayobj.py:1724]: I checked and for some reason, the source index array is always non-contiguous. So it's the other way around, the former branch (no_copy) isn't being hit.

File: numba/core/typing/arraydecl.py
Comment (kc611) [reply numba/core/typing/arraydecl.py:92]: Probably not, will add one.

File: numba/np/arrayobj.py
Comment (sklam) [inline numba/np/arrayobj.py:780]: Bad syntax

File: numba/np/arrayobj.py
Comment (kc611) [reply numba/np/arrayobj.py:1662]: With the current implementation, I don't think that's possible since fancy indexing always creates a copy of the data. i.e. In fancy indexing the source and destination array never share memory. This is stated in the NumPy documentation as well. 

I'm in-fact curious as of why it was the case over here that they did.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @kc611. I've given them a review and I think there's a few things to be resolved but otherwise looks good. The unit tests that are failing CI are failing due to the use of `compile_isolated` and `Flags()`. To fix, this:
https://github.com/numba/numba/blob/74cb715132519cbe5dcc4ea963c6953ef424861d/numba/tests/test_indexing.py#L15-L16
could have:
```
enable_pyobj_flags.nrt = True
```
adding. Or alternatively, as object mode fallback is deprecated, forced object mode testing could be achieved by using alternative flags such as those available through importing these from testing support:
https://github.com/numba/numba/blob/74cb715132519cbe5dcc4ea963c6953ef424861d/numba/tests/support.py#L70-L71
Hope this helps?

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:12]: dead code/imports?

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:490]:
```suggestion
             (0, np.array([[0, 1, 0, 1, 0, 1], [1, 0, 1, 0, 1, 0]], dtype=bool), slice(None), 3,))
```

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:314]:
```suggestion
        (3, np.array([[0, 1, 3, 4, 2], [0, 1, 2, 3, 2], [3, 1, 3, 4, 1]]),
```

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:320]:
```suggestion
        (np.array([[0, 1, 3, 4, 2], [0, 1, 2, 3, 2], [3, 1, 3, 4, 1]]),
```

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:327]:
```suggestion
        (np.array([[0, 1, 3, 4, 2], [0, 1, 2, 3, 2], [3, 1, 3, 4, 1]]),
         3, Ellipsis, slice(1, 5)), # multi-dimensional
```

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:341]:
```suggestion
        (Ellipsis, 1, np.array([0, 1, 3, 4, 2], order='A'), 3, slice(1, 5)),
        (np.array([0, 1, 3, 4, 2], order='F'), 3, Ellipsis, slice(1, 5)),
        (np.array([[0, 1, 3, 4, 2], [0, 1, 2, 3, 2], [3, 1, 3, 4, 1]], order='A'),
         3, Ellipsis, slice(1, 5)), # multidimensional
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @kc611, few minor things to resolve now else looks good. OOB I spoke with @gmarkall about the impact of the use of the NRT on the CUDA target. The conclusion is as follows:
1. It's probably best to concentrate on getting this and the next related PR (#8912) merged as the implementation needs to be figured out to start with and it's almost certainly easier to do this on the CPU with the NRT present.
2. Go back to e.g. the `release0.57` branch and see what sort of "fancy indexing" actually worked on CUDA, write tests for this.
3. Get the tests in 2. working on `main` once these PRs are merged. I suspect it'll only be the case where the index is a non-contiguous array that's a problem as the copy is currently needed as part of "flattening"  it. Once we have the tests I expect the options and limitations will become more clear.

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:763]: Ideally this would be tested, but I'm inclined to leave this to a subsequent PR as I am relatively sure it's going to need to be moved.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1680]: As above, ideally this would be tested, but I'm inclined to leave this to a subsequent PR as I am relatively sure it's going to need to be moved.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [reply numba/np/arrayobj.py:1171]: Resolving this can be postponed as the 3 points in https://github.com/numba/numba/pull/8491#pullrequestreview-1411169708 will cover it.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1743]: This needs moving to a module global in a subsequent PR.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1790]: Use of NRT needs assessing, defer to later PR.

File: numba/tests/test_fancy_indexing.py
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:337]: Setting `order='A'` in the NumPy constructor won't have the effect of it being `A` ordered within Numba's type system (assuming that was the intent?) Suggest deferring fixing this to a later PR.

**Review by stuartarchibald — APPROVED** [review]
Many thanks for all your efforts on this @kc611. It's great to see this feature implemented. There's a few outstanding issues that I've left as a "review", but these should be captured in a new issue for completion in subsequent work. Most of the outstanding items relate to use of the Numba runtime which is going to take careful assessment. As alluded to previously, this will manifest in having to write a number of tests for the CUDA target to ensure that no regressions are introduced. Thanks again for work on this challenging implementation and feature, patch is approved!

**Pull request merged**
